csstidy/csstidy.1
.\"     Title: csstidy
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets v1.72.0 <http://docbook.sf.net/>
.\"      Date: 03/18/2007
.\"    Manual: 
.\"    Source: 
.\"
.TH "CSSTIDY" "1" "03/18/2007" "" ""
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
csstidy \- CSS parser and optimiser
.SH "SYNOPSIS"
\fIcsstidy\fR input\-file [options] [output\-file]
.sp
.SH "DESCRIPTION"
The csstidy(1) command parses and optimises a css file given as \fIinput\-file\fR. If no \fIoutput\-file\fR is specified, the results are displayed to standard output. Various options allow for fine\-grained control over the level of readability, file size and other items.
.sp
If you use the filename "\-", csstidy reads from stdin. In most case you will probably find that csstidy defaults are reasonable and applicable to most situations.
.sp
Note that csstidy is not designed to be a css validator. Although you should only enter valid (or rather well\-formed) code, csstidy will also parse and fix incorrect code.
.sp
.SH "EXAMPLES"
.PP
To optimise a CSS file and send the output to screen use:
.RS 4
csstidy mycssfile.css
.RE
.PP
If you want to save the output to a file, use:
.RS 4
csstidy mycssfile.css myoutputfile.css
.RE
.PP
If you want to use an option[s] and output to a file, use:
.RS 4
csstidy mycssfile.css \-\-lowercase_s=true myoutputfile.css
.RE
.SH "OPTIONS"
.PP
\-\-allow_html_in_templates=[false|true]
.RS 4
default = false
.RE
.PP
\-\-case_properties=[false|true]
.RS 4
Case for properties. Default = false.
.RE
.PP
\-\-compress_colors=[true|false]
.RS 4
If set to true, csstidy will optimize color value by trying to save bytes (i.e. black becomes #000). Default = true.
.RE
.PP
\-\-compress_font\-weight=[true|false]
.RS 4
If set to true, csstidy will optimize font value by trying to save bytes (i.e. normal becomes 400). Default = true.
.RE
.PP
\-\-discard_invalid_properties=[false|true]
.RS 4
If set to true, csstidy will remove invalid properties. In a sense this is like validation, except you get no warnings unless you study the output. As an example, if you misspelled the property "width" as "with", that property will simply be removed. Default = false.
.RE
.PP
\-\-lowercase_s=[false|true]
.RS 4
If set to true, all uppercase selectors (i.e. A, P, SPAN) will be rewritten in lowercase. Default = false.
.RE
.PP
\-\-preserve_css=[false|true]
.RS 4
All comments and most browser specific hacks will be removed if this is set to false. Setting this to true will preserve browser specific hacks. Default = false.
.RE
.PP
\-\-strip_all_comments=[false|true]
.RS 4
Strip all comments even left on preserve_css set to true. Default = false.
.RE
.PP
\-\-remove_bslash=[true|false]
.RS 4
Remove unnecessary backslashes. This may or may not be desirable depending upon whether you have used backslashes to accomodate browsers that do not comply with w3c standards. Default = true.
.RE
.PP
\-\-remove_last_semicolon=[false|true]
.RS 4
Default = false.
.RE
.PP
\-\-silent=[false|true]
.RS 4
When set to true, csstidy will not output information about the optimizations preformed. Default = false.
.RE
.PP
\-\-sort_properties=[false|true]
.RS 4
Sorts properties (i.e. background, color, width) of selectors. Default = false.
.RE
.PP
\-\-sort_selectors=[false|true]
.RS 4
Sorts selectors. Use this with caution, as it will totally rearrange any sense of layout in your file that you may have written into it, and css is sensitive to order. Sorts in order of ID's, classes and then selectors. Default = false.
.RE
.PP
\-\-timestamp=[false|true]
.RS 4
If set to true, adds a comment enclosed timestamp to the top of the output. Default = false.
.RE
.PP
\-\-merge_selectors=[2|1|0]
.RS 4
If selectors (including ID's and classes) have identical properties, then csstidy will merge them. A setting of 2 represents merging, 1 won't merge and 0 means no change of already existing merges. Default = 2.
.RE
.PP
\-\-optimise_shorthands=[1|2|0]
.RS 4
If set to 2, csstidy will invoke all optimisations. 0 represents no optimisations, and 1 a safe level of optimatisations. Default = 1.
.RE
.PP
\-\-template=[default|filename|low|high|highest]
.RS 4
Affects how much whitespace is removed and ultimately the readability of the output.
.RE
.SH "CSS HACKS"
The following is a list of all CSS hacks that will not be modified by default by csstidy. If you enable "preserve CSS", more hacks will be saved. In general you should do without CSS hacks anyway, but if you need them and want to use csstidy you should use one of the following hacks:
.sp
.sp
.RS 4
.nf
p\eroperty:value; (only if you disable the removal of unnecessary backslashes)
div#test
head:first\-child+body div
body>div
html[xmlns] div
@import "null?\e"\e\e{"; @import "styles.css"; (and all other @import rules)
@media all{/\e* rules \e*/}
* html div
i\e{content:"\e"/*"\e} div\e{property:value\e}
html*#test
_property:value
.fi
.RE
.SH "AUTHOR"
Written by Florian Schmitz <floele@gmail.com>,
corrections by Frank W. Bergmann
.sp
.SH "RESOURCES"
Main web site: \fIhttp://csstidy.sourceforge.net/\fR
.sp
This manual page was written by Kevin Coyner <kcoyner@debian.org> for the Debian project but may be used by others.
.sp


NEWS
CSSTidy NEWS -- history of user-visible changes.  2006-05-19

Please send CSSTidy bug reports to http://sourceforge.net/tracker/?atid=771415&group_id=148404

Version 1.3
* SCons support added
* There is now a `NEWS' file (this one), giving a history of
  user-visible changes.



csstidy/csstidy.cpp
/*
 * Copyright 2005, 2006, 2007 Florian Schmitz
 *
 * This file is part of CSSTidy.
 *
 *  CSSTidy is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation; either version 2.1 of the License, or
 *   (at your option) any later version.
 *
 *   CSSTidy is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU Lesser General Public License for more details.
 * 
 *   You should have received a copy of the GNU Lesser General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
 
#include "csspp_globals.hpp"
extern map< string, vector<string> > shorthands;

csstidy::csstidy()
{ 
	properties = 0;
	selectors = 0;
	charset = "";
	namesp = "";
	line = 1;
	tokens = "{};:()@='\"/,\\!$%&*+.<>?[]^`|~";
	css_level = "CSS3.0";
	
	settings["remove_bslash"] = 1;
	settings["compress_colors"] = 1;
	settings["compress_font-weight"] = 1;
	settings["lowercase_s"] = 0;
	settings["optimise_shorthands"] = 1;
	settings["optimise_numbers"] = 0;
	settings["remove_last_semicolon"] = 0;
	settings["case_properties"] = 0;
	settings["sort_properties"] = 0;
	settings["sort_selectors"] = 0;
	settings["merge_selectors"] = 2;
	settings["discard_invalid_properties"] = 0;
	settings["allow_html_in_templates"] = 0;
	settings["silent"] = 0;
	settings["preserve_css"] = 0;
	settings["strip_all_comments"] = 0;
	settings["timestamp"] = 0;
	
	csstemplate.push_back("<span class=\"at\">"); //string before @rule
	csstemplate.push_back("</span> <span class=\"format\">{</span>\n"); //bracket after @-rule
	csstemplate.push_back("<span class=\"selector\">"); //string before selector
	csstemplate.push_back("</span> <span class=\"format\">{</span>\n"); //bracket after selector
	csstemplate.push_back("<span class=\"property\">"); //string before property
	csstemplate.push_back("</span><span class=\"value\">"); //string after property+before value
	csstemplate.push_back("</span><span class=\"format\">;</span>\n"); //string after value
	csstemplate.push_back("<span class=\"format\">}</span>"); //closing bracket - selector
	csstemplate.push_back("\n\n"); //space between blocks {...}
	csstemplate.push_back("\n<span class=\"format\">}</span>\n\n"); //closing bracket @-rule
	csstemplate.push_back(""); //indent in @-rule
	csstemplate.push_back("<span class=\"comment\">"); // before comment
	csstemplate.push_back("</span>\n"); //after comment
	csstemplate.push_back("\n"); // after last line @-rule
} 
	
void csstidy::add_token(const token_type ttype, const string data, const bool force)
{
	if(settings["strip_all_comments"] && (ttype == COMMENT)) return;
	if(settings["preserve_css"] || force) {
		token temp;
		temp.type = ttype;
		temp.data = (ttype == COMMENT) ? data : trim(data);
		csstokens.push_back(temp);
	}
}

void csstidy::copy(const string media, const string selector, const string media_new, const string selector_new)
{
	for(int k = 0; k < css[media][selector].size(); k++)
	{	
		string property = css[media][selector].at(k);
		string value = css[media][selector][property];
		add(media_new,selector_new,property,value);
	}
}

void csstidy::add(const string& media, const string& selector, const string& property, const string& value)
{
	if(settings["preserve_css"]) {
		return;
	}
	
	if(css[media][selector].has(property))
	{
		if( !is_important(css[media][selector][property]) || (is_important(css[media][selector][property]) && is_important(value)) )
		{
			css[media][selector].erase(property);
			css[media][selector][property] = trim(value);
		}
	}
	else
	{
		css[media][selector][property] = trim(value);
	}
}

void csstidy::log(const string msg, const message_type type, int iline)
{
	message new_msg;
	new_msg.m = msg;
	new_msg.t = type;
	if(iline == 0)
	{
		iline = line;
	}
	if(logs.count(line) > 0)
	{
		for(int i = 0; i < logs[line].size(); ++i)
		{
			if(logs[line][i].m == new_msg.m && logs[line][i].t == new_msg.t)
			{
				return;
			}
		}
	}
	logs[line].push_back(new_msg);
}

string csstidy::unicode(string& istring,int& i)
{
	++i;
	string add = "";
	bool replaced = false;
	
	while(i < istring.length() && (ctype_xdigit(istring[i]) || ctype_space(istring[i])) && add.length()< 6)
	{
		add += istring[i];

		if(ctype_space(istring[i]))
		{
			break;
		}
		i++;
	}

	if(hexdec(add) > 47 && hexdec(add) < 58 || hexdec(add) > 64 && hexdec(add) < 91 || hexdec(add) > 96 && hexdec(add) < 123)
	{
		string msg = "Replaced unicode notation: Changed \\" + rtrim(add) + " to ";
		add = static_cast<int>(hexdec(add));
		msg += add;
		log(msg,Information);
		replaced = true;
	}
	else
	{
		add = trim("\\" + add);
	}

	if(ctype_xdigit(istring[i+1]) && ctype_space(istring[i]) && !replaced || !ctype_space(istring[i]))
	{
		i--;
	}
	
	if(add != "\\" || !settings["remove_bslash"] || in_str_array(tokens,istring[i+1]))
	{
		return add;
	}
	if(add == "\\")
	{
		log("Removed unnecessary backslash",Information);
	}
	return "";
}

bool csstidy::is_token(string& istring,const int i)
{
	return (in_str_array(tokens,istring[i]) && !escaped(istring,i));
}

void csstidy::merge_4value_shorthands(string media, string selector)
{
	for(map< string, vector<string> >::iterator i = shorthands.begin(); i != shorthands.end(); ++i )
	{
		string temp;

		if(css[media][selector].has(i->second[0]) && css[media][selector].has(i->second[1])
		&& css[media][selector].has(i->second[2]) && css[media][selector].has(i->second[3]))
		{
			string important = "";
			for(int j = 0; j < 4; ++j)
			{
				string val = css[media][selector][i->second[j]];
				if(is_important(val))
				{
					important = "!important";
					temp += gvw_important(val)+ " ";
				}
				else
				{
					temp += val + " ";
				}
				css[media][selector].erase(i->second[j]);
			}
			add(media, selector, i->first, shorthand(trim(temp + important)));		
		}
	}
} 

map<string,string> csstidy::dissolve_4value_shorthands(string property, string value)
{
	map<string, string> ret;
	extern map< string, vector<string> > shorthands;
	
	if(shorthands[property][0] == "0")
	{
		ret[property] = value;
		return ret;
	}
	
	string important = "";
	if(is_important(value))
	{
		value = gvw_important(value);
		important = "!important";
	}
	vector<string> values = explode(" ",value);

	if(values.size() == 4)
	{
		for(int i=0; i < 4; ++i)
		{
			ret[shorthands[property][i]] = values[i] + important;
		}
	}
	else if(values.size() == 3)
	{
		ret[shorthands[property][0]] = values[0] + important;
		ret[shorthands[property][1]] = values[1] + important;
		ret[shorthands[property][3]] = values[1] + important;
		ret[shorthands[property][2]] = values[2] + important;
	}
	else if(values.size() == 2)
	{
		for(int i = 0; i < 4; ++i)
		{
			ret[shorthands[property][i]] = ((i % 2 != 0)) ? values[1] + important : values[0] + important;
		}
	}
	else
	{
		for(int i = 0; i < 4; ++i)
		{
			ret[shorthands[property][i]] = values[0] + important;
		}	
	}
	
	return ret;
}

void csstidy::explode_selectors()
{
	// Explode multiple selectors
    if (settings["merge_selectors"] == 1)
    {
        vector<string> new_sels;
        int lastpos = 0;
        sel_separate.push_back(cur_selector.length());
        
        for (int i = 0; i < sel_separate.size(); ++i)
        {
            if (i == sel_separate.size()-1) {
                sel_separate[i] += 1;
            }
            
            new_sels.push_back(cur_selector.substr(lastpos,sel_separate[i]-lastpos-1));
            lastpos = sel_separate[i];
        }
 
        if (new_sels.size() > 1)
        {
            for (int i = 0; i < new_sels.size(); ++i)
            {
				for (pstore::iterator j = css[cur_at][cur_selector].begin(); j != css[cur_at][cur_selector].end(); ++j)
				{
            		add(cur_at, new_sels[i], j->first, j->second);
				}
            }
            css[cur_at].erase(cur_selector);
        }
    }
    sel_separate = vector<int>();
}
